### ISSUE TITLE
Confusing features in dataset returned by `get_torch_loader`

### ISSUE THREAD

**Issue by eswan01** [pinder-org/pinder#20]
I'm trying to use the dataloaders returned by `pinder.core.loader.dataset.get_torch_loader` for all-atom representations of the protein complexes. When I iterate over the keys present in a sample from one of these datasets, I get:
```
from pinder.core.loader.dataset import PinderDataset, get_torch_loader
train, val = [get_torch_loader(PinderDataset(split=split)) for split in ["train", "val"]]
for key in train.dataset[0].keys():
    print(key, type(train.dataset[0][key]))
    if type(train.dataset[0][key]) == dict:
        for key2 in train.dataset[0][key].keys():
            print("    ", key2, type(train.dataset[0][key][key2]), train.dataset[0][key][key2].shape)

### ------- Output ------- ###

target_complex <class 'dict'>
     atom_types <class 'torch.Tensor'> torch.Size([2556, 12])
     residue_types <class 'torch.Tensor'> torch.Size([2556, 1])
     atom_coordinates <class 'torch.Tensor'> torch.Size([2556, 3])
     residue_coordinates <class 'torch.Tensor'> torch.Size([2556, 3])
     residue_ids <class 'torch.Tensor'> torch.Size([2556])
     chain_ids <class 'torch.Tensor'> torch.Size([2556])
feature_complex <class 'dict'>
     atom_types <class 'torch.Tensor'> torch.Size([2556, 12])
     residue_types <class 'torch.Tensor'> torch.Size([2556, 1])
     atom_coordinates <class 'torch.Tensor'> torch.Size([2556, 3])
     residue_coordinates <class 'torch.Tensor'> torch.Size([2556, 3])
     residue_ids <class 'torch.Tensor'> torch.Size([2556])
     chain_ids <class 'torch.Tensor'> torch.Size([2556])
id <class 'str'>
sample_id <class 'str'>
target_id <class 'str'>
```
I have the following questions:

**What is the difference between `atom_coordinates` and `residue_coordinates`?**

For at least the first 10 examples in the train dataset, I find that
```
for i in range(10):
    print((train.dataset[i]['feature_complex']['residue_coordinates'] == train.dataset[i]['feature_complex']['atom_coordinates']).all())
```
returns `True`.

**Can the loader be modified to also return atom name information?**

The `atom_types` tensor provides only a one-hot encoding of element type, for example, running
```
train.dataset[0]['target_complex']['atom_types'].unique(dim=0)
```
returns
```
tensor([[0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])
```
This is confusing when compared to the `Structure` class, in which `atom_name` is encoded as elements of `constants.BB_ATOMS` and `constants.SC_ATOMS` (e.g. 'N', 'CA', 'C', 'O', 'CB', 'CG', etc.).

**It would be useful/less confusing if the dataloader returned by `get_torch_loader` contained information that corresponded more closely to the data provided in the `Structure` class.**

**Comment by danielkovtun** [comment]
Hi @eswan01, thanks for the feedback on the example transform implementation. I agree that it would make sense to rename the misleading `atom_types` key to reflect that its actually encodings of elements and not atom types and also introduce an atom type encoding. 


## Regarding your first question

> What is the difference between atom_coordinates and residue_coordinates?

In the `PinderDataset` (torch dataset), there is no difference. It appears this was an oversight when hijacking the existing transform that was targeting the torch-geometric dataset. In the torch-geometric dataset, the difference is that `residue_coordinates` tracks only the C-alpha atom coordinates (one per residue) and `atom_coordinates` tracks all-atom. Note the `receptor.filter("atom_name", mask=["CA"])` below:
https://github.com/pinder-org/pinder/blob/bb62635bcb4186cf4f45696fc3da5082bb899f0a/src/pinder-core/pinder/core/loader/geodata.py#L118-L133

That `structure2tensor` function really just serves as an example transform for how one could go from `Structure` instances to Tensor-encoded data items. In the `PinderDataset` implementation, the _default_  torch-API transform defined is the `structure2tensor_transform` which essentially just wraps the `structure2tensor` method such that it accepts a `Structure` as input and returns a dict of tensors. 
https://github.com/pinder-org/pinder/blob/bb62635bcb4186cf4f45696fc3da5082bb899f0a/src/pinder-core/pinder/core/loader/dataset.py#L238-L254

As you can see below, the current version uses all-atom coordinates for both `residue_coordinates` and `atom_coordinates` (the oversight):
https://github.com/pinder-org/pinder/blob/bb62635bcb4186cf4f45696fc3da5082bb899f0a/src/pinder-core/pinder/core/loader/dataset.py#L54-L63

I will push a fix to that default transform shortly 👍 


## Regarding the second question

> Can the loader be modified to also return atom name information?

As I alluded to above, we can definitely adjust the info being returned by the default transform as you suggested. I will try to include that in the upcoming patch for the `residue_coordinates` bug. 

That being said, the short answer to 

> Can the loader be modified to also return atom name information? 
> It would be useful/less confusing if the dataloader returned by get_torch_loader contained information that corresponded more closely to the data provided in the Structure class.
> 

**Absolutely!** You can define whatever transform you desire, that's why the `PinderLoader` default return value of `__getitem__` contains a tuple of `(PinderSystem, Structure(<feature complex>), Structure(<target complex>))` 

https://github.com/pinder-org/pinder/blob/bb62635bcb4186cf4f45696fc3da5082bb899f0a/src/pinder-core/pinder/core/loader/dataset.py#L293-L299

You can specify your own `transform` and `target_transform` to apply to the `feature_complex` `Structure` and `target_complex` `Structure`. This closely follows the transform/target_transform concepts from the `torch` datasets + dataloader tutorial: https://pytorch.org/tutorials/beginner/basics/data_tutorial.html#creating-a-custom-dataset-for-your-files
